libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp

```diff
@@ -467,38 +468,91 @@ IPAddress ESP8266WiFiSTAClass::dnsIP(uint8_t dns_no) {
+   name. No distinction is made between upper and lower case.  The first
+   character must be an alpha character.  The last character must not be
+   a minus sign or period.  A host which serves as a GATEWAY should have
+   "-GATEWAY" or "-GW" as part of its name.  Hosts which do not serve as
+   Internet gateways should not use "-GATEWAY" and "-GW" as part of
+   their names. A host which is a TAC should have "-TAC" as the last
+   part of its host name, if it is a DoD host.  Single character names
+   or nicknames are not allowed.
+  ^^^^ RFC952 ^^^^
+
+  - 24 chars max
+  - only a..z A..Z 0..9 -]
+  - no '-' as last char
+  */
+
+    if (aHostname.length() == 0) {
+        DEBUGV("WiFi.(set)hostname(): empty name\n");
         return false;
     }
-    return wifi_station_set_hostname(aHostname);
-}
 
-/**
- * Set ESP8266 station DHCP hostname
- * @param aHostname max length:32
- * @return ok
- */
-bool ESP8266WiFiSTAClass::hostname(const char* aHostname) {
-    return hostname((char*) aHostname);
-}
+    // rework hostname to be RFC compliant
 
-/**
- * Set ESP8266 station DHCP hostname
- * @param aHostname max length:32
- * @return ok
- */
-bool ESP8266WiFiSTAClass::hostname(const String& aHostname) {
-    return hostname((char*) aHostname.c_str());
+    if (aHostname.length() > 24) {
+        // nonos-sdk allows 32, but
+        // dhcp server may require RFC compliance
+        aHostname.remove(24);
+    }
+    for (size_t i = 0; i < aHostname.length(); i++)
+        // replace unallowed chars by dashes
+        if (!isalnum(aHostname[i]) && aHostname[i] != '-')
+            aHostname[i] = '-';
+    if (aHostname[aHostname.length() - 1] == '-') {
+        // check for ending dash
+        aHostname[aHostname.length() - 1] = 'x';
+    }
+
+    bool ret = wifi_station_set_hostname(aHostname.c_str());
+
+    if (!ret) {
+#ifdef DEBUG_ESP_CORE
+        static const char fmt[] PROGMEM = "WiFi.hostname(%s): wifi_station_set_hostname() failed\n";
+        DEBUGV(fmt, aHostname.c_str());
+#endif
+        return false;
+    }
+
+    // now we should inform dhcp server for this change, using lwip_renew()
+    // looping through all existing interface
+    // harmless for AP, also compatible with ethernet adapters (to come)
+    for (netif* intf = netif_list; intf; intf = intf->next) {
+
+        // unconditionally update all known interfaces
+        intf->hostname = wifi_station_get_hostname();
+
+        if (netif_dhcp_data(intf) != nullptr) {
+            // renew already started DHCP leases
+            err_t lwipret = dhcp_renew(intf);
+            if (lwipret != ERR_OK) {
+#ifdef DEBUG_ESP_CORE
+                static const char fmt[] PROGMEM = "WiFi.hostname(%s): lwIP error %d on interface %c%c (index %d)\n";
+                DEBUGV(fmt, intf->hostname, (int)lwipret, intf->name[0], intf->name[1], intf->num);
+#endif
+                ret = false;
+            }
+        }

```

Comment:
How about an else here to cover the case of static IP setup, where dhcp_inform() is called?
Does dhcp_inform() work instead of dhcp_renew() in the dhcp case?

Reply:
We can't, dhcp servers have absolutely no knowledge on static IPs originating from devices.
On static leases however, this PRs still applies.
This dhcp option "hostname" is not supposed to modify DNSs. It works when your DNS and DHCP servers talk to each other (like dnsmasq offering both services). Otherwise, DNS servers cannot be aware of these dhcp names.

Reply:
> dhcp servers have absolutely no knowledge on static IPs originating from devices 
My understanding, both from a quick overview of the dhcp inform message, and of the dhcp_inform() lwip function comment, is that this is precisely the reason for sending the message: it lets local dhcp servers know that there is a static IP in use.
A search online didn't give me much details about how dhcp_inform() is supposed to work, but I did find one hint about this use case, and it's as you said: the default behavior for dhcp servers is to ignore all inform messages. Therefore, this isn't useful.